src/crimson/osd/pg_recovery.cc

```diff
@@ -0,0 +1,466 @@
+	soid);
+      started += pg->prep_object_replica_pushes(
+	soid, r->second.need, out);
+    }
+  }
+
+  return started;
+}
+
+template<typename T>
+size_t PGRecovery<T>::start_backfill_ops(
+  size_t max_to_start,
+  std::vector<crimson::osd::blocking_future<>> *out)
+{
+  T* pg = static_cast<T*>(this);
+  crimson::get_logger(ceph_subsys_osd).debug(
+    "{}({}): bft={} lbs={} {}",
+    __func__,
+    pg->peering_state.get_backfill_targets(),
+    last_backfill_started,
+    new_backfill ? "" : "new_backfill");
+  assert(!pg->peering_state.get_backfill_targets().empty());
+
+  ceph_abort("not implemented!");
+}
+
+template<typename T>
+std::vector<pg_shard_t> PGRecovery<T>::get_replica_recovery_order() const
+{
+  const T* pg = static_cast<const T*>(this);
+  std::vector<std::pair<unsigned int, pg_shard_t>> replicas_by_num_missing,
+    async_by_num_missing;
+  replicas_by_num_missing.reserve(
+    pg->peering_state.get_acting_recovery_backfill().size() - 1);
+  for (auto &p: pg->peering_state.get_acting_recovery_backfill()) {
+    if (p == pg->peering_state.get_primary()) {
+      continue;
+    }
+    auto pm = pg->peering_state.get_peer_missing().find(p);
+    assert(pm != pg->peering_state.get_peer_missing().end());
+    auto nm = pm->second.num_missing();
+    if (nm != 0) {
+      if (pg->peering_state.is_async_recovery_target(p)) {
+	async_by_num_missing.push_back(make_pair(nm, p));
+      } else {
+	replicas_by_num_missing.push_back(make_pair(nm, p));
+      }
+    }
+  }
+  // sort by number of missing objects, in ascending order.
+  auto func = [](const std::pair<unsigned int, pg_shard_t> &lhs,
+		 const std::pair<unsigned int, pg_shard_t> &rhs) {
+    return lhs.first < rhs.first;
+  };
+  // acting goes first
+  std::sort(replicas_by_num_missing.begin(), replicas_by_num_missing.end(), func);
+  // then async_recovery_targets
+  std::sort(async_by_num_missing.begin(), async_by_num_missing.end(), func);
+  replicas_by_num_missing.insert(replicas_by_num_missing.end(),
+    async_by_num_missing.begin(), async_by_num_missing.end());
+
+  std::vector<pg_shard_t> ret;
+  ret.reserve(replicas_by_num_missing.size());
+  for (auto p : replicas_by_num_missing) {
+    ret.push_back(p.second);
+  }
+  return ret;
+}
+
+template<typename T>
+std::optional<crimson::osd::blocking_future<>> PGRecovery<T>::recover_missing(
+  const hobject_t &soid, eversion_t need)
+{
+  T* pg = static_cast<T*>(this);
+  if (pg->peering_state.get_missing_loc().is_deleted(soid)) {
+    return pg->recovery_backend->get_recovering(soid).make_blocking_future(
+	pg->recovery_backend->recover_delete(soid, need));
+  } else {
+    return pg->recovery_backend->get_recovering(soid).make_blocking_future(
+      pg->recovery_backend->recover_object(soid, need).handle_exception(
+	[=, soid = std::move(soid)] (auto e) {
+	on_failed_recover({ pg->get_pg_whoami() }, soid, need);
+	return seastar::make_ready_future<>();
+      })
+    );
+  }
+}
+
+template<typename T>
```

Comment:
Um...only PG for now

Reply:
do you mean I should change this to PGRecovery<PG>::XXX?

Reply:
My view is that: a template class instead of a class increases the complexity. If we don't need it (nor envision we'll need in the future!), then let's keep the things simple.

Reply:
You mean, instead of implementing the PGRecovery as a base class for PG, we can implement it as a normal class with a reference to PG in it and add a reference to PGRecovery to PG, is this right?:-)

Reply:
Right. CRTP is pretty confusing pattern. If we can avoid it with a simpler approach, I would opt for that.

Reply:
Normally, CRTP is used to avoid the overhead of dynamic dispatch.  Perfectly ok as far as that goes.  However, I don't think the dynamic dispatch cost here will be detectable at all as I don't think any of these interface boundaries fall within a tight loop.

Moreover, we would ideally like to be able to create unittests for this interface which don't involve an actual PG.  Thus, I think we should:
1) Let PGRecovery be its own class with a PGRecoveryListener (or a better name) interface implemented by PG and (potentially) by a test fixture).
2) Let the PGRecovery instance be a member of PG and simply pass it a pointer to PG as an implementer of the PGRecoveryListener (or w/e) interface.  The pointer overhead involved is negligible.

Reply:
Got it:-)